### PR TITLE
Support user-defined set types

### LINKED ISSUE #23: Let user select set-type used in tracers
While internal use of `BitSet` for the index sets showed a large performance increase on the Brusselator example in #4, it struggles with very large input dimensionalities, as `BitSet` is [designed for use with dense integer sets](https://docs.julialang.org/en/v1/base/collections/#Base.BitSet).

For very large, very sparse index sets, `Set{UInt64}` or implementations from other packages like [SmallCollections.jl](https://github.com/matthias314/SmallCollections.jl) might be better suited.

Ideally, users could select the right set type for their task. 

CC: @gdalle, @niklasschmitz

### PR DESCRIPTION
**Pull request by adrhill**
Closes #23 by supporting all sets of type `AbstractSet{<:Integer}`.
Breaking since the current `pattern` API now requires the user to specify the set type in the tracer:
```Julia-repl
julia> using SparseConnectivityTracer

julia> x = rand(3);

julia> f(x) = [x[1]^2, 2 * x[1] * x[2]^2, sin(x[3])];

julia> pattern(f, JacobianTracer{BitSet}, x) # <--- Set type needs to be specifies
3×3 SparseArrays.SparseMatrixCSC{Bool, UInt64} with 4 stored entries:
 1  ⋅  ⋅
 1  1  ⋅
 ⋅  ⋅  1
 ```

### CODE REVIEW

File: test/references/show/ConnectivityTracer.txt [hunk 1]
```diff
@@ -1 +1 @@
-ConnectivityTracer(1, 2, 3)
\ No newline at end of file
+ConnectivityTracer{BitSet}(2,)
```
Comment (gdalle) [inline test/references/show/ConnectivityTracer.txt:1]: why has this changed?

File: test/references/show/HessianTracer.txt [hunk 2]
```diff
@@ -1,5 +1,3 @@
-HessianTracer(
+HessianTracer{BitSet}(
   2 => (),
-  3 => (),
-  1 => (),
```
Comment (gdalle) [inline test/references/show/HessianTracer.txt:4]: why has this changed?

File: test/references/show/JacobianTracer.txt [hunk 3]
```diff
@@ -1 +1 @@
-JacobianTracer(1, 2, 3)
\ No newline at end of file
+JacobianTracer{BitSet}(2,)
```
Comment (gdalle) [inline test/references/show/JacobianTracer.txt:1]: why has this changed?

File: test/runtests.jl [hunk 4]
```diff
@@ -61,85 +61,86 @@ DocMeta.setdocmeta!(
...
+
+        t = tracer(ConnectivityTracer{BitSet}, 2)
         @test ConnectivityTracer(t) == t
-        @test empty(t) == empty(ConnectivityTracer)
-        @test ConnectivityTracer(1) == empty(ConnectivityTracer)
+        @test empty(t) == empty(ConnectivityTracer{BitSet})
+        @test ConnectivityTracer{BitSet}(1) == empty(ConnectivityTracer{BitSet})
 
-        t = tracer(JacobianTracer, 1, 2, 3)
+        t = tracer(JacobianTracer{BitSet}, 2)
```
Comment (gdalle) [inline test/runtests.jl:103]: why has this changed?

File: test/runtests.jl [hunk 5]
```diff
@@ -61,85 +61,86 @@ DocMeta.setdocmeta!(
...
         @test JacobianTracer(t) == t
-        @test empty(t) == empty(JacobianTracer)
-        @test JacobianTracer(1) == empty(JacobianTracer)
+        @test empty(t) == empty(JacobianTracer{BitSet})
+        @test JacobianTracer{BitSet}(1) == empty(JacobianTracer{BitSet})
 
         # Base.show
         @test_reference "references/show/ConnectivityTracer.txt" repr(
-            "text/plain", tracer(ConnectivityTracer, 1, 2, 3)
+            "text/plain", tracer(ConnectivityTracer{BitSet}, 2)
```
Comment (gdalle) [inline test/runtests.jl:110]: why has this changed?

File: test/runtests.jl [hunk 6]
```diff
@@ -61,85 +61,86 @@ DocMeta.setdocmeta!(
...
         )
         @test_reference "references/show/JacobianTracer.txt" repr(
-            "text/plain", tracer(JacobianTracer, 1, 2, 3)
+            "text/plain", tracer(JacobianTracer{BitSet}, 2)
```
Comment (gdalle) [inline test/runtests.jl:113]: why has this changed?

File: test/runtests.jl [hunk 7]
```diff
@@ -257,16 +258,16 @@ DocMeta.setdocmeta!(
 
         # Base.show
         @test_reference "references/show/HessianTracer.txt" repr(
-            "text/plain", tracer(HessianTracer, 1, 2, 3)
+            "text/plain", tracer(HessianTracer{BitSet}, 2)
```
Comment (gdalle) [inline test/runtests.jl:261]: why has this changed?

**Review by gdalle — COMMENTED** [review]
You were right it's too complex for me to review so I just flagged a few things that stood out to me as abnormal

What is the point of these reference tests if the reference files are changed automatically?

File: test/references/show/ConnectivityTracer.txt (hunk 1 above)
Comment (adrhill) [reply test/references/show/ConnectivityTracer.txt]: I added it for code coverage of `Base.show` 🤷‍♂️

**Pull request merged**